packages/@aws-cdk/assets/lib/fs/utils.ts

```diff
@@ -58,3 +59,112 @@ export function shouldFollow(mode: FollowMode, sourceRoot: string, realPath: str
     return path.resolve(realPath).startsWith(path.resolve(sourceRoot));
   }
 }
+
+type AssetFile = {
+  absolutePath: string;
+  relativePath: string;
+  isFile: boolean;
+  isDirectory: boolean;
+  size: number;
+} & ({ isSymbolicLink: false } | { isSymbolicLink: true; symlinkTarget: string });
+
+const generateAssetFile = (rootDir: string, fullFilePath: string, stat: fs.Stats): AssetFile => ({
+  absolutePath: fullFilePath,
+  relativePath: path.relative(rootDir, fullFilePath),
+  isFile: stat.isFile(),
+  isDirectory: stat.isDirectory(),
+  size: stat.size,
+  isSymbolicLink: false,
+});
+
+const generateAssetSymlinkFile = (rootDir: string, fullFilePath: string, stat: fs.Stats, symlinkTarget: string): AssetFile => ({
+  ...generateAssetFile(rootDir, fullFilePath, stat),
+  isSymbolicLink: true,
+  symlinkTarget,
+});
+
+export function listFilesRecursively(dir: string, options: CopyOptions & Required<Pick<CopyOptions, 'follow'>>, _rootDir?: string): AssetFile[] {
+  const files: AssetFile[] = [];
+  let exclude = options.exclude || [];
+  const rootDir = _rootDir || dir;
+  const followStatsFn = options.follow === FollowMode.ALWAYS ? fs.statSync : fs.lstatSync;
+
+  recurse(dir);
+
+  function recurse(currentPath: string, currentStat?: fs.Stats): void {
+    {
+      const stat = currentStat || fs.statSync(currentPath);
+      if (!stat) {
+        return;
+      }
+
+      if (!stat.isDirectory()) {
+        files.push(generateAssetFile(rootDir, currentPath, stat));
+        return;
+      }
+    }
+
+    for (const file of fs.readdirSync(currentPath)) {
+      const fullFilePath = path.join(currentPath, file);
+
+      let stat: fs.Stats | undefined = followStatsFn(fullFilePath);
+      if (!stat) {
+        continue;
+      }
+
+      let target = '';
+      if (stat.isSymbolicLink()) {
+        target = fs.readlinkSync(fullFilePath);
+
+        // determine if this is an external link (i.e. the target's absolute path
+        // is outside of the root directory).
+        const targetPath = path.normalize(path.resolve(currentPath, target));
+
+        if (shouldFollow(options.follow, rootDir, targetPath)) {
+          stat = fs.statSync(fullFilePath);
+          if (!stat) {
+            continue;
+          }
+        }
+      }
+
+      const relativeFilePath = path.relative(rootDir, fullFilePath);
+
+      // we've just discovered that we have a directory
+      if (stat.isDirectory()) {
+        // to help future shouldExclude calls, we're changing the exlusion patterns
+        // by expliciting "dir" exclusions to "dir/*" (same with "!dir" -> "!dir/*")
+        exclude = exclude.reduce<string[]>((res, pattern) => {
+          res.push(pattern);
+          if (pattern.trim().replace(/^!/, '') === relativeFilePath) {
```

Comment:
Does not look like this will correctly handle cases where the directory entry in the ignore list contains a wildcard.

Reply:
Can you give me an example? Because I think I'm already testing this use-case:
https://github.com/aws/aws-cdk/blob/baa431539a0b024d9f065cfed8a27e24771ce6cc/packages/%40aws-cdk/aws-ecr-assets/test/test.image-asset.ts#L339